### ISSUE TITLE
torch arange with dtype=bfloat16 causes inaccuracies

### ISSUE THREAD

**Issue by p-sodmann** [pytorch/pytorch#114544]
### 🐛 Describe the bug

When creating an enumerated tensor using torch.arrange with dtype=torch.bfloat16, there may be a one-off error in the result.

enumerate_tensor = torch.arange(start = 0, end=2048, step=1, device=device, dtype=torch.bfloat16) 
enumerate_tensor[1001] --> tensor(1000., device='cuda:0', dtype=torch.bfloat16)

enumerate_tensor = torch.arange(start = 0, end=2048, step=1, device=device) 
enumerate_tensor[1001] --> tensor(1001, device='cuda:0')

### Versions

PyTorch version: 2.1.0+cu121
Is debug build: False
CUDA used to build PyTorch: 12.1
ROCM used to build PyTorch: N/A

OS: Microsoft Windows 11 Pro for Workstations
GCC version: Could not collect
Clang version: Could not collect
CMake version: Could not collect
Libc version: N/A

Python version: 3.11.3 | packaged by Anaconda, Inc. | (main, Apr 19 2023, 23:46:34) [MSC v.1916 64 bit (AMD64)] (64-bit runtime)
Python platform: Windows-10-10.0.22621-SP0
Is CUDA available: True
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: NVIDIA GeForce RTX 4090
Nvidia driver version: 536.67
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture=9
CurrentClockSpeed=3000
DeviceID=CPU0
Family=207
L2CacheSize=32768
L2CacheSpeed=
Manufacturer=GenuineIntel
MaxClockSpeed=3000
Name=13th Gen Intel(R) Core(TM) i9-13900K
ProcessorType=3
Revision=

Versions of relevant libraries:
[pip3] adabelief-pytorch==0.2.1
[pip3] denoising-diffusion-pytorch==1.8.11
[pip3] efficientnet-pytorch==0.7.1
[pip3] ema-pytorch==0.2.3
[pip3] flake8==6.0.0
[pip3] memory-efficient-attention-pytorch==0.1.6
[pip3] mypy-extensions==0.4.3
[pip3] numpy==1.25.0
[pip3] numpydoc==1.5.0
[pip3] pytorch-fid==0.3.0
[pip3] pytorch-lightning==2.1.2
[pip3] pytorch-toolbelt==0.6.3
[pip3] segmentation-models-pytorch==0.3.3
[pip3] torch==2.1.0+cu121
[pip3] torchaudio==2.1.0+cu121
[pip3] torchmetriclogger==0.1
[pip3] torchmetrics==1.0.3
[pip3] torchsde==0.2.5
[pip3] torchvision==0.16.0+cu121
[conda] adabelief-pytorch         0.2.1                    pypi_0    pypi
[conda] blas                      1.0                         mkl
[conda] denoising-diffusion-pytorch 1.8.11                   pypi_0    pypi
[conda] efficientnet-pytorch      0.7.1                    pypi_0    pypi
[conda] ema-pytorch               0.2.3                    pypi_0    pypi
[conda] memory-efficient-attention-pytorch 0.1.6                    pypi_0    pypi
[conda] mkl                       2023.1.0         h8bd8f75_46356
[conda] mkl-service               2.4.0           py311h2bbff1b_1
[conda] mkl_fft                   1.3.6           py311hf62ec03_1
[conda] mkl_random                1.2.2           py311hf62ec03_1
[conda] numpy                     1.25.0                   pypi_0    pypi
[conda] numpydoc                  1.5.0           py311haa95532_0
[conda] pytorch-fid               0.3.0                    pypi_0    pypi
[conda] pytorch-lightning         2.1.2                    pypi_0    pypi
[conda] pytorch-toolbelt          0.6.3                    pypi_0    pypi
[conda] segmentation-models-pytorch 0.3.3                    pypi_0    pypi
[conda] torch                     2.1.0+cu121              pypi_0    pypi
[conda] torchaudio                2.1.0+cu121              pypi_0    pypi
[conda] torchmetriclogger         0.1                      pypi_0    pypi
[conda] torchmetrics              1.0.3                    pypi_0    pypi
[conda] torchsde                  0.2.5                    pypi_0    pypi
[conda] torchvision               0.16.0+cu121             pypi_0    pypi

**Comment by jon-chuang** [comment]
I think this might be a bfloat16 accuracy thing. It can represent up to 2-3 decimals of precision, so I'm guessing that the 1 in 1001 gets truncated. But not sure, maybe there's a better way of generating bfloat16 ranges.

**Comment by malfet** [comment]
Yes, this is an expected behavior, as bfloat16 has only 7 bit fraction and 1001 needs 10 bits:
```
 % python
Python 3.9.15 | packaged by conda-forge | (main, Nov 22 2022, 08:48:25) 
[Clang 14.0.6 ] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> int(1001).bit_length()
10
>>> bin(1001)
'0b1111101001'
>>> import torch
torch.tensor([>>> torch.tensor([1001.0], dtype=torch.bfloat16), torch.tensor([1001.0], dtype=torch.float16)
(tensor([1000.], dtype=torch.bfloat16), tensor([1001.], dtype=torch.float16))
```
